test/infrastructure/docker/internal/provisioning/cloudinit/writefiles.go

```diff
@@ -62,13 +66,79 @@ func newWriteFilesAction() action {
 	return &writeFilesAction{}
 }
 
-func (a *writeFilesAction) Unmarshal(userData []byte) error {
+func (a *writeFilesAction) Unmarshal(userData []byte, kindMapping kind.Mapping) error {
 	if err := yaml.Unmarshal(userData, a); err != nil {
 		return errors.Wrapf(err, "error parsing write_files action: %s", userData)
 	}
+	for i, f := range a.Files {
+		if f.Path == kubeadmInitPath {
+			// NOTE: in case of join the kubeadmConfigFile contains both the ClusterConfiguration and the InitConfiguration
+			contentSplit := strings.Split(f.Content, "---\n")
+
+			if len(contentSplit) != 3 {
+				return errors.Errorf("invalid kubeadm config file, unable to parse it")
+			}
+			initConfiguration, err := kubeadmtypes.UnmarshalInitConfiguration(contentSplit[2])
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse init configuration")
+			}
+
+			fixNodeRegistration(&initConfiguration.NodeRegistration, kindMapping.Mode)
+
+			contentSplit[2], err = kubeadmtypes.MarshalInitConfigurationForVersion(initConfiguration, kindMapping.KubernetesVersion)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse init configuration")
+			}
+			a.Files[i].Content = strings.Join(contentSplit, "---\n")
+		}
+		if f.Path == kubeadmJoinPath {
+			// NOTE: in case of join the kubeadmConfigFile contains only the join Configuration
+			joinConfiguration, err := kubeadmtypes.UnmarshalJoinConfiguration(f.Content)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse join configuration")
+			}
+
+			fixNodeRegistration(&joinConfiguration.NodeRegistration, kindMapping.Mode)
+
+			a.Files[i].Content, err = kubeadmtypes.MarshalJoinConfigurationForVersion(joinConfiguration, kindMapping.KubernetesVersion)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse join configuration")
+			}
+		}
+	}
 	return nil
 }
 
+// fixNodeRegistration sets node registration for running Kubernetes/kubelet in docker.
+// NOTE: we add those values if they do not exists; user can set those flags to different values to disable automatic fixing.
```

Comment:
But there is no way for users to unset the values right?

I think it's fine to set args that we know are pretty safe to set, but I wonder about stuff like eviction-hard. I have no idea why it is necessary to set it and if I see correctly now they don't have a way to unset it.

Reply:
It is possible to disable defaulting by setting parameters to the default value, but ultimately I'm not really concerned about this, because this PR makes using CAPD simpler in CI and in the majority of use cases for local development.
Unsetting those flags will be an exception, and it is acceptable to have a slightly better use case for exceptions.

Reply:
What I was trying to say, maybe we shouldn't hard-code all flags, just the once where we know that they are necessary for things to work.

Because if they lead to problems folks cannot just unset them in the cluster-template they have to fork CAPD>

Reply:
I'm assuming what we have in templates is required, do you have different evidence?
Generally speaking, I think we should optimize for the use case we know, and as I said there is a way around by setting the default value for the flag.
But if you are over concerned about this I can also add something on release notes and we can change course when we have more concreate use cases to reason about

Reply:
Alright, let's go with that for now. We can change it if we run into problems.